devdoc/Metrics.rst

```diff
@@ -0,0 +1,282 @@
+The framework defines a sink interface to support the different requirements of a collection systems.
+Examples of collection systems are Prometheus, Datadog, and Elasticsearch. Each has different requirements
+for how and when measurements are ingested. The following are examples of different collection system
+requirements:
+
+* Polled vs Periodic
+* Single measurement vs multiple reports
+* Report format (JSON, text, etc.)
+* Push vs Pull
+
+Sinks are responsible for two main functions: initiating a collection and reporting
+measurements to the collection system. The *Metrics Reporter* provides the support to complete
+these functions.
+
+The sink encapsulates all of the collection system requirements providing a pluggable architecture that
+isolates components from these differences. The framework supports multiple sinks concurrently,
+each operating independently.
+
+Instrumented components are not aware of the sink or sinks in use. Sinks can be changed without
+requiring changes to a component. Therefore, components are independent of the collection system(s) in use.
+
+
+Metrics Reporter
+================
+
+The metrics reporter class provides all of the common functions to bind together the component, the metrics it creates, and the sinks to which measurements are reported. It is responsible for the following:
+
+* Initialization of the framework
+* Managing the metrics created by the component
+* Handling collection and reporting as directed by configured sinks
+
+
+Metrics Implementations
+=======================
+
+The sections that follow discuss metric implementations.
+
+Gauge Metric
+------------
+A gauge metric is a continuously updated value representing the current state of an interesting value in the component. For example, the amount of memory used in an internal buffer, or the number of requests waiting on a queue. A gauge metric may increase or decrease in value as needed. Reading the value of a gauge is a stateless operation in that there are no dependencies on the previous reading. The value returned shall always be the current state.
+
+Once created, the component shall update the gauge anytime the state of what is measured is updated. The metric shall provide methods to increase and decrease the value. The sink reads the value during collection an reporting.
+
+
+*************
+Configuration
+*************
+This section discusses configuration. Since Helm charts are capable of combining configuration data at a global level into a component's specific configuration, a single config source is all all that is required. The configuration takes the form as shown below. Note that as the design progresses it is expected that there will be additions.
+
+::
+
+  component:
+    metrics:
+      sinks:
+        - name: <sink name>
+          type: <sink_type>
+          settings:
+            sink_setting1: sink_setting_value1
+            sink_setting2: sink_setting_value2
+          metrics:
+            - name: <metric_name>
```

Comment:
worth having a real example to illustrate the structure.  Especially how the structure supports options being set on a component to override a global setting.

Reply:
Need to better understand how helm accomplishes this. The YML presented in the doc is assumed to be the combined config from both global and component pieces and represents what the framework parses.

Reply:
That is a good idea to show the merged config, but we also need the user config to provide some documentation, and ensure the structure is correct.

Reply:
To summarise the discussion.  For the first version we only have a global metric configuration and all components that support metrics use that same configuration.  We revisit once there is a requirement to configure a sink on a per component basis.  (And then merge either in the helm generation or quite possibly within the framework code.)